### PR TITLE
zmq4: resend subscriptions in socket.addConn

### PR DESCRIPTION
**Pull request by thielepaul**
see #116

### CODE REVIEW

File: socket.go [hunk 1]
```diff
@@ -30,12 +30,13 @@ var (
...
-	retry time.Duration
-	sec   Security
-	log   *log.Logger
+	ep       string // socket end-point
+	typ      SocketType
+	id       SocketIdentity
+	retry    time.Duration
+	sec      Security
+	log      *log.Logger
+	topicsCb func() []string
```
Comment (sbinet) [inline socket.go:39]:
```suggestion
	topics func() []string
```

also, could you add a test for this?

**Review by sbinet — COMMENTED** [review]
looking good.
see my little nit-pick oriented comments below.

(thanks for the PR)

File: socket.go [hunk 2]
```diff
@@ -286,6 +287,12 @@ func (sck *socket) addConn(c *Conn) {
 	if sck.r != nil {
 		sck.r.addConn(c)
 	}
+	// resend subscriptions for topics if there are any
+	if sck.subTopics != nil {
+		for _, topic := range sck.subTopics() {
+			_ = sck.Send(NewMsg(append([]byte{1}, topic...)))
+		}
+	}
 	sck.mu.Unlock()
```
Comment (sbinet) [inline socket.go:296]: perhaps move this up as a `defer sck.mu.Unlock()`

File: socket_test.go [hunk 3]
```diff
@@ -220,3 +221,37 @@ func TestConnReaperDeadlock(t *testing.T) {
...
+		if err != nil {
+			return
+		}
+	}()
+	msg, err := sub.Recv()
+	if err != nil {
+		t.Errorf("Recv failed: %v", err)
+	}
+	if string(msg.Frames[0]) != "test" {
+		t.Errorf("Expected 'test' got '%s'", msg.String())
```
Comment (sbinet) [inline socket_test.go:255]: the Go idiom is to print "got" before "want"

File: socket_test.go [hunk 4]
```diff
@@ -220,3 +221,37 @@ func TestConnReaperDeadlock(t *testing.T) {
...
+		defer wg.Done()
+		time.Sleep(1 * time.Millisecond)
+		err := pub.Send(zmq4.NewMsgFromString([]string{"test"}))
+		if err != nil {
+			return
+		}
+	}()
+	msg, err := sub.Recv()
+	if err != nil {
+		t.Errorf("Recv failed: %v", err)
```
Comment (sbinet) [inline socket_test.go:252]:
```suggestion
		t.Failf("Recv failed: %v", err)
```

File: socket_test.go [hunk 5]
```diff
@@ -220,3 +221,37 @@ func TestConnReaperDeadlock(t *testing.T) {
...
+func TestSocketSendSubscriptionOnConnect(t *testing.T) {
+	endpoint := "inproc://test-resub"
+
+	sub := zmq4.NewSub(context.Background())
+	defer sub.Close()
+	pub := zmq4.NewPub(context.Background())
+	defer pub.Close()
+	sub.SetOption(zmq4.OptionSubscribe, "test")
+	if err := sub.Listen(endpoint); err != nil {
+		t.Errorf("Sub Dial failed: %v", err)
```
Comment (sbinet) [inline socket_test.go:234]:
```suggestion
		t.Failf("Sub Dial failed: %v", err)
```
(no point in going through with the rest of the test)

File: socket_test.go [hunk 6]
```diff
@@ -220,3 +221,37 @@ func TestConnReaperDeadlock(t *testing.T) {
...
+	}
+	if err := pub.Dial(endpoint); err != nil {
+		t.Errorf("Pub Dial failed: %v", err)
```
Comment (sbinet) [inline socket_test.go:237]:
```suggestion
		t.Failf("Pub Dial failed: %v", err)
```
(no point in going through with the rest of the test)

File: socket.go [hunk 7]
```diff
@@ -286,6 +287,12 @@ func (sck *socket) addConn(c *Conn) {
 	if sck.r != nil {
 		sck.r.addConn(c)
 	}
+	// resend subscriptions for topics if there are any
+	if sck.topicsCb != nil {
+		for _, topic := range sck.topicsCb() {
+			sck.Send(NewMsg(append([]byte{1}, topic...)))
```
[the remark this answers is not in the document]
Comment (thielepaul) [reply socket.go]: done

File: socket.go (hunk 1 above)
Comment (thielepaul) [reply socket.go]: `topics()` is already used for the pub topics, I renamed it to `subTopics`

File: socket.go (hunk 2 above)
Comment (thielepaul) [reply socket.go]: done

File: socket_test.go (hunk 3 above)
Comment (thielepaul) [reply socket_test.go]: done

File: socket_test.go (hunk 4 above)
Comment (thielepaul) [reply socket_test.go]: I guess you mean `t.Fatalf` ?

File: socket_test.go (hunk 6 above)
Comment (thielepaul) [reply socket_test.go]: I guess you mean `t.Fatalf` ?

File: socket_test.go (hunk 5 above)
Comment (thielepaul) [reply socket_test.go]: I guess you mean `t.Fatalf` ?

**Comment by thielepaul** [comment]
@sbinet I don't understand how the coverage in `conn.go` is would be reduced by my changes, do you understand what is going on there?

**Comment by sbinet** [comment]
code coverage (with codecov) can be a bit flaky, in my experience.

File: socket_test.go (hunk 4 above)
Comment (sbinet) [reply socket_test.go]: yes, sorry.

**Review by sbinet — APPROVED** [review]
LGTM.

could you send another PR against [go-zeromq/license](https://github.com/go-zeromq/license), adding yourself to the `AUTHORS` and/or `CONTRIBUTORS` files?

thanks again for the PR.

**Comment by sbinet** [comment]
needs https://github.com/go-zeromq/license/pull/12

**Comment by sbinet** [comment]
thanks again (and welcome!)

**Pull request merged**